### PR TITLE
Enable setting the StopCaptureReason

### PR DESCRIPTION
**Pull request by vickyliu-go4it**
With this change, we allow the `StartStopCaptureRequestWaiter` to set
the `StopCaptureReason`. This allows the CloudCollector to call
`StopCapture` with different reasons such as `kExceedsTimeDuration`,
`kUploaderStop`, `kSignalStop` and so on.

Bug: http://b/223124163

### CODE REVIEW

File: src/CaptureServiceBase/include/CaptureServiceBase/StartStopCaptureRequestWaiter.h [hunk 1]
```diff
@@ -17,6 +18,16 @@ class StartStopCaptureRequestWaiter {
...
+
+  void SetStopCaptureReason(CaptureServiceBase::StopCaptureReason stop_capture_reason) {
+    stop_capture_reason_ = std::move(stop_capture_reason);
+  }
+  [[nodiscard]] std::optional<CaptureServiceBase::StopCaptureReason> GetStopCaptureReason() const {
+    return stop_capture_reason_;
+  }
+
+ private:
+  std::optional<CaptureServiceBase::StopCaptureReason> stop_capture_reason_;
```
Comment (dpallotti) [inline src/CaptureServiceBase/include/CaptureServiceBase/StartStopCaptureRequestWaiter.h:30]: Probably as expected, `StartStopCaptureRequestWaiter` is getting more complicated than originally intended. For example, it's no longer an interface. I'm missing the context quite a bit, in particular around `CloudCollectorStartStopCaptureRequestWaiter` and the structure of the cloud collector, so I'm not sure if there's clearly a better way really. Maybe @beckerhe has more thoughts.

File: src/LinuxCaptureService/LinuxCaptureServiceBase.cpp [hunk 2]
```diff
@@ -205,18 +207,16 @@ LinuxCaptureServiceBase::WaitForStopCaptureRequestOrMemoryThresholdExceeded(
       ORBIT_LOG("Memory threshold exceeded: stopping capture (and stopping memory watchdog)");
       absl::MutexLock lock{stop_capture_mutex.get()};
       *stop_capture = true;
-      *stop_capture_reason = StopCaptureReason::kMemoryWatchdog;
+      if (!start_stop_capture_request_waiter->GetStopCaptureReason().has_value()) {
+        start_stop_capture_request_waiter->SetStopCaptureReason(StopCaptureReason::kMemoryWatchdog);
+      }
```
[the remark this answers is not in the document]
Comment (vickyliu-go4it) [reply src/LinuxCaptureService/LinuxCaptureServiceBase.cpp:212]: Thanks for pointing this out! I will add a mutex to the `StartStopCaptureRequestWaiter` to guard the `stop_capture_reason_`.

**Comment by vickyliu-go4it** [comment]
@dpallotti Hi Davide, could you help to take a look at this new implementation?

We changed `StartStopCaptureRequestWaiter::WaitForStopCaptureRequest` to return an external `StopCaptureReason`. For the `GrpcStartStopCaptureRequestWaiter`, this is always `kClientStop`; and for the `CloudCollectorStartStopCaptureRequestWaiter`, we require the caller of `StopCapture` to provide a `StopCaptureReason` (and it is thread safe as the `stop_capture_reason_` is guarded by `CloudCollectorStartStopCaptureRequestWaiter::stop_mutex_`).

**Review by dpallotti — APPROVED** [review]
Nice, much cleaner.

**Review by beckerhe — APPROVED** [review]
Indeed much nicer. Thank you Vicky!

**Pull request merged**